packages/components/src/image_zoom/hooks.jsx

```diff
@@ -0,0 +1,233 @@
+} from './utils';
+import { CONTAINER_WIDTH_CHANGE_RATE, MAX_SCALE_FACTOR } from './constants';
+
+export const usePrevious = (value) => {
+  const ref = useRef(null);
+  useEffect(() => { ref.current = value; });
+
+  return ref.current;
+};
+
+export const useImage = (src) => {
+  const [image, setImage] = useState(null);
+
+  useEffect(() => {
+    const imageObj = new Image(100, 100); // width and height get ignored
+    imageObj.src = src;
+    imageObj.onload = () => { setImage(imageObj); };
+  }, [src]);
+
+  return image;
+};
+
+const useContainerWidth = (ref) => {
+  const [width, setWidth] = useState(getElementWidth(ref?.current));
+
+  const handleResize = useDebouncedCallback(() => {
+    if (!ref.current) return;
+
+    setWidth(getElementWidth(ref.current));
+  }, CONTAINER_WIDTH_CHANGE_RATE);
+
+  useEffect(() => {
+    const el = ref.current;
+    if (!el) return;
+
+    handleResize();
+
+    const resizeObserver = new ResizeObserver(() => handleResize());
+    resizeObserver.observe(el);
+
+    return () => { resizeObserver.disconnect(); };
+  }, [ref.current, handleResize]);
+
+  return width;
+};
+
+export const usePreviewSize = (rootRef, aspectRatio) => {
+  const rootWidth = useContainerWidth(rootRef);
+
+  return useMemo(() => {
+    if (rootWidth === undefined) {
+      return { width: 0, height: 0 };
+    }
+
+    return {
+      width: rootWidth,
+      height: rootWidth * (aspectRatio ** -1),
+    };
+  }, [rootWidth]);
+};
+
+export const usePinchZoom = (onAnchorZoom) => {
+  const pinchZoomRef = useRef(null);
+
+  const start = (pointA, pointB) => {
+    const lastDistance = getDistanceBetweenPoints(pointA, pointB);
+    pinchZoomRef.current = { lastDistance };
+  };
+
+  const move = (pointA, pointB) => {
+    const distance = getDistanceBetweenPoints(pointA, pointB);
+    const midpoint = getMidpoint(pointA, pointB);
+
+    const zoomFactor = distance / pinchZoomRef.current.lastDistance;
+
+    onAnchorZoom(zoomFactor, midpoint);
+
+    pinchZoomRef.current.lastDistance = distance;
+  };
+
+  const stop = () => {
+    pinchZoomRef.current = null;
+  };
+
+  return { start, stop, move };
+};
+
+export const useDrag = (onUpdate) => {
+  const dragRef = useRef(null);
+
+  const start = (position, offset) => {
+    const { x: offsetX, y: offsetY } = position;
+    const { top: startTop, left: startLeft } = offset;
+
+    dragRef.current = { offsetX, offsetY, startTop, startLeft };
+  };
+
+  const move = (position) => {
+    const origin = dragRef.current;
+    if (!origin) return;
+
+    const { x: offsetX, y: offsetY } = position;
+
+    onUpdate(
+      getOffsetForMovement(origin, { offsetX, offsetY }),
+    );
+  };
+
+  const stop = () => {
+    dragRef.current = null;
+  };
+
+  return { start, move, stop };
+};
+
+const TYPE_RESET = 'reset';
+const TYPE_ANCHOR_ZOOM = 'anchor-zoom';
+const TYPE_SAFE_SET_OFFSET = 'safe-set-offset';
+
+export const useImageView = ({ previewSize, imageSize }) => {
```

Comment:
Open for suggestions how to further refactor this. It's important to use a reducer here because many state variables depend on one another